File: python/p0206.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val  # The value of the node
        self.next = next  # Pointer to the next node in the list

class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        # If the list is empty, return None
        if not head:
            return None

        # If the list has only one node, return that node
        if not head.next:
            return head

        prev_node = None  # This will become the new tail (initially None)
        curr_node = head  # Start with the head of the list

        # Iterate through the list until we reach the end
        while curr_node:
            next_node = curr_node.next  # Save reference to the next node
            curr_node.next = prev_node  # Reverse the current node's pointer
            prev_node = curr_node  # Update previous node to be the current node
            curr_node = next_node  # Move to the next node

        # The loop ends when curr_node is None (end of the list)
        # prev_node is now the new head of the reversed list
        return prev_node

# Example usage:
# Creating a linked list: 1 -> 2 -> 3 -> 4 -> 5 -> None
node1 = ListNode(1)
node2 = ListNode(2)
node3 = ListNode(3)
node4 = ListNode(4)
node5 = ListNode(5)

node1.next = node2
node2.next = node3
node3.next = node4
node4.next = node5

# Before reversal, the list is: 1 -> 2 -> 3 -> 4 -> 5 -> None
solution = Solution()
reversed_head = solution.reverseList(node1)

# After reversal, the list should be: 5 -> 4 -> 3 -> 2 -> 1 -> None
# Let's print the reversed list to verify
current = reversed_head
while current:
    print(current.val, end=" -> " if current.next else " -> None\n")
    current = current.next

File: javascript/p0206.js
// Definition for singly-linked list.
function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}

/**
 * Reverses a singly-linked list.
 * @param {ListNode} head - The head of the singly-linked list.
 * @return {ListNode} - The new head of the reversed list.
 */
var reverseList = function (head) {
  let prev_node = null;
  let cur_node = head;

  while (cur_node) {
    let next_node = cur_node.next; // Save reference to the next node
    cur_node.next = prev_node; // Reverse the current node's pointer
    prev_node = cur_node; // Move prev_node to the current node
    cur_node = next_node; // Move to the next node
  }

  return prev_node; // Return the new head of the reversed list
};

// Example usage:
// Creating a linked list: 1 -> 2 -> 3 -> 4 -> 5 -> null
let node1 = new ListNode(1);
let node2 = new ListNode(2);
let node3 = new ListNode(3);
let node4 = new ListNode(4);
let node5 = new ListNode(5);

node1.next = node2;
node2.next = node3;
node3.next = node4;
node4.next = node5;

// Before reversal, the list is: 1 -> 2 -> 3 -> 4 -> 5 -> null
let reversedHead = reverseList(node1);

// After reversal, the list should be: 5 -> 4 -> 3 -> 2 -> 1 -> null
// Let's print the reversed list to verify
let current = reversedHead;
while (current) {
  process.stdout.write(current.val + (current.next ? " -> " : " -> null\n"));
  current = current.next;
}
